src/main/java/org/assertj/core/internal/Iterables.java

```diff
@@ -453,16 +453,62 @@ public void assertContainsOnlyNulls(AssertionInfo info, Iterable<?> actual) {
    * @throws AssertionError if the given {@code Iterable} does not contain the given sequence of objects.
    */
   public void assertContainsSequence(AssertionInfo info, Iterable<?> actual, Object[] sequence) {
-    if (commonCheckThatIterableAssertionSucceeds(info, actual, sequence)) return;
-    // check for elements in values that are missing in actual.
-    List<?> actualAsList = newArrayList(actual);
-    for (int i = 0; i < actualAsList.size(); i++) {
-      // look for given sequence in actual starting from current index (i)
-      if (containsSequenceAtGivenIndex(actualAsList, sequence, i)) return;
+    // perform the checks that would have been done in commonCheckThatIterableAssertionSucceeds
+    // but do them explicitly without having to create a new iterator on actual - which would
+    // break if actual were only singly-traversable.
+    checkNotNullIterables(info, actual, sequence);
+    // store the elements from actual that have been visited (because we don't know we can look ahead - the 'actual'
+    // might be singly-traversable) in a cyclic list, which can then be rolled for each new element
+    // until a match is found or until the 'actual' is exhausted. Of course if 'actual' really is infinite then this could
+    // take a while :-D
+    // once it's full, check if it equals sequence for each new element
+    RingBuffer buffer = new RingBuffer(sequence.length);
```

Comment:
Move the RingBuffer creation after the call to `failIfEmptySinceActualIsNotEmpty` as we don't need until the while loop.